### PR TITLE
CI/CD Improvements

### PR DESCRIPTION
**Pull request by JacobMGEvans**
### Tests run before other workflows
To be sure that Tests and type checking is occurring on code before it is merged and released other workflows have the Testing workflow as a trigger. For example the Tests need to complete and be successful before the prerelease will run.

### Release flow triggered on PR's closed
Whenever the a PR is closed/merged the Workflow will trigger for checking for a changeset and adding it to the changeset "Version" PR, when the "Version" PR is closed/merged it will trigger a release.
 ### Consolidated old workflows as jobs in new workflows
Combined the old workflows with new workflows that have similar behavior & shared similar event triggers.

### CODE REVIEW

**Comment by JacobMGEvans** [comment]
@petebacondarwin

File: .github/workflows/release.yml
Comment (petebacondarwin) [inline .github/workflows/release.yml:7]: Do we want to use the same strategy of triggering off commits to master?

File: .github/workflows/release.yml
Comment (petebacondarwin) [inline .github/workflows/release.yml:30]: In which case this would change.
Also it would be good to ensure that this only runs on the cloudflare owner.

File: .github/workflows/typechecks.yml
Comment (petebacondarwin) [inline .github/workflows/typechecks.yml:9]:
```suggestion
  pull_request:
```

Comment (petebacondarwin) [reply .github/workflows/typechecks.yml:9]: To ensure this runs when new commits are added.

File: .github/workflows/typechecks.yml
Comment (petebacondarwin) [inline .github/workflows/typechecks.yml:22]:
```suggestion
      - name: Setup Node.js ${{ matrix.node_version }}
```

File: .github/workflows/typechecks.yml
Comment (petebacondarwin) [inline .github/workflows/typechecks.yml:51]: Remove??

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
I think we can piggy back on a lot of the learning from the wrangler2 project from the last few days.

**Comment by threepointone** [comment]
We also have to make changes to the way the automated types are generated internally to include a changeset.

File: .github/workflows/typechecks.yml
Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml:51]: Yes!

File: .github/workflows/release.yml
Comment (JacobMGEvans) [reply .github/workflows/release.yml:30]: Done.

File: .github/workflows/typechecks.yml
Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml:9]: Done

File: .github/workflows/typechecks.yml
Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml:22]: Currently dont have a matrix setup, or know what Node version are supported. This is just the Wrangler supported Node version.

File: .github/workflows/typechecks.yml
Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml:51]: Done 😄

**Comment by JacobMGEvans** [comment]
> We also have to make changes to the way the automated types are generated internally to include a changeset.

I will make the necessary changes for the current changes, however, I will hold off on the PR then until we can talk more about this.

File: .github/workflows/typechecks.yml
Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml:20]: I can add something like 
`strategy:
      matrix:
        node-version: [16.x, 17.x, 18.x]`

**Comment by JacobMGEvans** [comment]
> We also have to make changes to the way the automated types are generated internally to include a changeset.

@threepointone There are two possible ways to do this from what I can tell.

1. So I have looked at the code in the edgeworker that does the "publish" (PR creation)  it uses Octokit, I can figure out how to use that API to include a changeset. Might involve adding changesets creation to step 2, where the files are added/generated

2. Create a GH Action that runs on these particular PRs to create a changeset using NPX.

File: .github/workflows/release.yml
Comment (threepointone) [inline .github/workflows/release.yml:55]:
```suggestion

      - name: Check Docs
        run: npm run export:docs
      - name: Check Overrides
        run: npm run export:overrides

```

File: .github/workflows/release.yml
Comment (threepointone) [inline .github/workflows/release.yml:28]:
```suggestion
```

File: .github/workflows/release.yml
Comment (threepointone) [inline .github/workflows/release.yml:31]:
```suggestion
```

File: .github/workflows/typechecks.yml
Comment (threepointone) [inline .github/workflows/typechecks.yml:19]: add `markdownlint` into `devDependencies` in `package.json` and add an npm script that calls `markdownlint '**/*.md' --ignore node_modules`, and reference that script name here.

File: .github/workflows/typechecks.yml
Comment (threepointone) [inline .github/workflows/typechecks.yml:40]: Didn't we land land on using this instead? https://github.com/cloudflare/wrangler2/blob/main/.github/workflows/prereleases.yml#L24-L30

File: .changeset/config.json
Comment (threepointone) [inline .changeset/config.json:10]: add a newline here

File: .github/workflows/typechecks.yml
Comment (threepointone) [inline .github/workflows/typechecks.yml:20]: why is this called test if it only does typechecking?

**Review by threepointone — CHANGES_REQUESTED** [review]
Dropped some suggestions

File: .github/workflows/typechecks.yml
Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml:20]: No reason I can think of, can change it to `typecheck`

Comment (petebacondarwin) [reply .github/workflows/typechecks.yml:20]: Well it is "testing" whether the types are valid... 😄

Comment (threepointone) [reply .github/workflows/typechecks.yml:20]: yeah I can buy that. maybe this should stay as `test`.

File: .github/workflows/typechecks.yml
Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml:40]: Yes, we did, I missed updating this one.

File: .changeset/config.json
Comment (JacobMGEvans) [reply .changeset/config.json:10]: I will check the Wrangler2 one too, its auto generated and might be missing from there as well.

File: .github/workflows/typechecks.yml
Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml]: The markdownlint-cli being installed as a `devDependencies` and called from the script was failing (locally it worked) inside of the GH action it was unable to find it. I checked  markdownlint-cli and markdownlint docs/README the examples looked liked this current setup for CI and was globally installed for local use. 
I alternatively got it to work inside the NPM script with NPX which I can also just put directly in the GH Action.
It does mean that it doesn't get cached with the other dependencies.

Comment (threepointone) [reply .github/workflows/typechecks.yml]: We should figure out why. Not a fan of using `npx` for this.

Comment (threepointone) [reply .github/workflows/typechecks.yml]: An alternate solution is to just not use this, and use `prettier --check` instead, I assume it does the same thing?

Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml]: > We should figure out why. Not a fan of using `npx` for this.

Just curious why? 

> An alternate solution is to just not use this, and use `prettier --check` instead, I assume it does the same thing?

I can look into this, I think your likely correct Prettier probably can handle this.

Comment (threepointone) [reply .github/workflows/typechecks.yml]: > Just curious why not a fan of npx for this?
Because it doesn't solve the actual problem? We miss out on versioning, it doesn't get cached, etc.

Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml]: I tried Prettier out, it seems to work fine locally but it does make a lot of initial changes (especially CHANGELOG.md) mainly due to Prettier's default line length (80 chars) and adding some spaces. It does make the code examples in Markdown all uniform which a bit of a bonus.

Comment (threepointone) [reply .github/workflows/typechecks.yml]: re: line length https://prettier.io/docs/en/options.html#prose-wrap

Comment (JacobMGEvans) [reply .github/workflows/typechecks.yml]: > re: line length https://prettier.io/docs/en/options.html#prose-wrap

I can turn it off here: https://github.com/cloudflare/workers-types/blob/master/.prettierrc#L3 or pass it into the CLI should minimize the initial Prettier impact and remove the markdownlinter dep

**Comment by JacobMGEvans** [comment]
>We also have to make changes to the way the automated types are generated internally to include a changeset.

Internal PR merged for changesets

File: package.json
Comment (threepointone) [inline package.json:16]: Weren't we going to replace this with prettier?

Comment (JacobMGEvans) [reply package.json]: I just pushed the recent changes for that. It includes the configurations to impact the existing Markdown as minimally as possible.

File: .github/workflows/typechecks.yml
Comment (threepointone) [inline .github/workflows/typechecks.yml:31]: The npm install and caching steps need to be added to the lint job above as well. Alternately move the lint section into the test section.

File: package.json
Comment (threepointone) [inline package.json:16]: This should be `prettier --check`, so it fails on problems. Not `prettier --write`, which will always pass successfully. 

(Sorry for the intermittent reviews, reviewing from phone, and still not feeling great 😕. Thank you for doing this)

Comment (threepointone) [reply package.json:16]: Also, is the explicit --parser flag necessary? Won't it automatically choose that? (I dunno, hence asking)

Comment (JacobMGEvans) [reply package.json:16]: My mistake, I thought this would just format the changes in CI before merging. 

(No worries, these things can always wait until you feel better.)

File: .github/workflows/release.yml
Comment (threepointone) [inline .github/workflows/release.yml:41]: Should we call prettier check here as well

Comment (JacobMGEvans) [reply .github/workflows/release.yml:41]: Yes, there could be cases that it catches changes that PR workflow was bypassed.

File: package.json
Comment (threepointone) [inline package.json:17]:
```suggestion
    "prettier:check": "prettier --check --parser markdown '**/*.md' --ignore-path ./LICENSE.md",
    "prettier": "prettier --write --parser markdown '**/*.md' --ignore-path ./LICENSE.md",
```
Why do we ignore the LICENSE file? And why shouldn't we run prettier on all files?

**Review by threepointone — COMMENTED** [review]
I think we need to recommend adding the prettier extension, and format on save, so people don't get frustrated when they try to send a PR. Copy relevant bits from https://github.com/cloudflare/wrangler2/tree/main/.vscode?

File: package.json
Comment (JacobMGEvans) [reply package.json:17]: I was trying to minimize initial changes, I can run everything. 

I was ignoring the License file because it didn't seem necessary.

**Comment by JacobMGEvans** [comment]
>  And why shouldn't we run prettier on all files?

I am gonna set up Prettier to run on all of the files 👍🏻

File: .changeset/config.json
Comment (JacobMGEvans) [inline .changeset/config.json:6]: Needs to be `public` access & changelog needs to be updated

**Review by threepointone — APPROVED** [review]
I *think* this looks good, so let's land it and see.

**Pull request merged**
